Duplicate identifier 'Rx'.

**TypeScript Version:**
2.0.3

**Code**
```
tsc --target es5 --allowJs dist/bundle.es2015.js --out dist/bundle.es5.js
```

**Expected behavior:**
Success

**Actual behavior:** 
```
> tsc --target es5 --allowJs dist/bundle.es2015.js --out dist/bundle.es5.js

dist/bundle.es2015.js(64740,22): error TS2300: Duplicate identifier 'Rx'.
dist/bundle.es2015.js(64759,22): error TS2300: Duplicate identifier 'Rx'.
```

Those line above looks like this in the file. I marked the line
```
/* tslint:disable:no-unused-variable */
// Subject imported before Observable to bypass circular dependency issue since
// Subject extends Observable and Observable references Subject in it's
// definition
/* tslint:enable:no-unused-variable */
/**
 * @typedef {Object} Rx.Scheduler // line 64740
 * @property {Scheduler} queue Schedules on a queue in the current event frame
 * (trampoline scheduler). Use this for iteration operations.
 * @property {Scheduler} asap Schedules on the micro task queue, which uses the
 * fastest transport mechanism available, either Node.js' `process.nextTick()`
 * or Web Worker MessageChannel or setTimeout or others. Use this for
 * asynchronous conversions.
 * @property {Scheduler} async Schedules work with `setInterval`. Use this for
 * time-based operations.
 * @property {Scheduler} animationFrame Schedules work with `requestAnimationFrame`.
 * Use this for synchronizing with the platform's painting
 */
let Scheduler = {
    asap,
    queue,
    animationFrame,
    async
};
/**
 * @typedef {Object} Rx.Symbol // line 64759
 * @property {Symbol|string} rxSubscriber A symbol to use as a property name to
 * retrieve an "Rx safe" Observer from an object. "Rx safety" can be defined as
 * an object that has all of the traits of an Rx Subscriber, including the
 * ability to add and remove subscriptions to the subscription chain and
 * guarantees involving event triggering (can't "next" after unsubscription,
 * etc).
 * @property {Symbol|string} observable A symbol to use as a property name to
 * retrieve an Observable as defined by the [ECMAScript "Observable" spec](https://github.com/zenparsing/es-observable).
 * @property {Symbol|string} iterator The ES6 symbol to use as a property name
 * to retrieve an iterator from an object.
 */
let Symbol$3 = {
    rxSubscriber: $$rxSubscriber,
    observable: $$observable,
    iterator: $$iterator
};
```
First, build my typescript to es2015
```
tsc -p tsconfig-es2015.json
```
Second, use rollup to roll all the es2015 js files into one
```
rollup -f iife -c -o dist/bundle.es2015.js
```

Maybe the rollup step screwed up.
I was following http://blog.mgechev.com/2016/06/26/tree-shaking-angular2-production-build-rollup-javascript/

--- a/src/compiler/binder.ts
+++ b/src/compiler/binder.ts
@@ -429,7 +429,7 @@ namespace ts {
                 //       during global merging in the checker. Why? The only case when ambient module is permitted inside another module is module augmentation
                 //       and this case is specially handled. Module augmentations should only be merged with original module definition
                 //       and should never be merged directly with other augmentation, and the latter case would be possible if automatic merge is allowed.
-                if (!isAmbientModule(node) && (hasExportModifier || container.flags & NodeFlags.ExportContext)) {
+                if ((!isAmbientModule(node) && (hasExportModifier || container.flags & NodeFlags.ExportContext)) || isJSDocTypedefNamespaceMember()) {
                     const exportKind =
                         (symbolFlags & SymbolFlags.Value ? SymbolFlags.ExportValue : 0) |
                         (symbolFlags & SymbolFlags.Type ? SymbolFlags.ExportType : 0) |
@@ -442,6 +442,13 @@ namespace ts {
                 else {
                     return declareSymbol(container.locals, undefined, node, symbolFlags, symbolExcludes);
                 }
+            }
+
+            function isJSDocTypedefNamespaceMember() {
+                return node.kind === SyntaxKind.JSDocTypedefTag &&
+                    container.kind === SyntaxKind.ModuleDeclaration &&
+                    (<JSDocTypedefTag>node).name &&
+                    container === (<JSDocTypedefTag>node).name.parent;
             }
         }
 
@@ -1787,6 +1794,18 @@ namespace ts {
             switch (node.kind) {
                 /* Strict mode checks */
                 case SyntaxKind.Identifier:
+                    // for typedef type names with namespaces
+                    if (isInJavaScriptFile(node) && node.parent && node.parent.kind === SyntaxKind.ModuleDeclaration) {
+                        let parentNode = node.parent;
+                        while (parentNode && parentNode.kind !== SyntaxKind.JSDocTypedefTag) {
+                            parentNode = parentNode.parent;
+                        }
+                        if (parentNode && parentNode.kind === SyntaxKind.JSDocTypedefTag && (<JSDocTypedefTag>parentNode).name === node) {
+                            bindBlockScopedDeclaration(<Declaration>parentNode, SymbolFlags.TypeAlias, SymbolFlags.TypeAliasExcludes);
+                            break;
+                        }
+                    }
+
                 case SyntaxKind.ThisKeyword:
                     if (currentFlow && (isExpression(node) || parent.kind === SyntaxKind.ShorthandPropertyAssignment)) {
                         node.flowNode = currentFlow;
@@ -1910,6 +1929,10 @@ namespace ts {
                 case SyntaxKind.InterfaceDeclaration:
                     return bindBlockScopedDeclaration(<Declaration>node, SymbolFlags.Interface, SymbolFlags.InterfaceExcludes);
                 case SyntaxKind.JSDocTypedefTag:
+                    if (!(<JSDocTypedefTag>node).fullName || (<JSDocTypedefTag>node).fullName.kind === SyntaxKind.Identifier) {
+                        return bindBlockScopedDeclaration(<Declaration>node, SymbolFlags.TypeAlias, SymbolFlags.TypeAliasExcludes);
+                    }
+                    break;
                 case SyntaxKind.TypeAliasDeclaration:
                     return bindBlockScopedDeclaration(<Declaration>node, SymbolFlags.TypeAlias, SymbolFlags.TypeAliasExcludes);
                 case SyntaxKind.EnumDeclaration:

--- a/src/compiler/binder.ts
+++ b/src/compiler/binder.ts
@@ -56,7 +56,7 @@ namespace ts {
         }
         // Only jsdoc typedef definition can exist in jsdoc namespace, and it should
         // be considered the same as type alias
-        else if (node.kind === SyntaxKind.Identifier && node.flags & NodeFlags.InJSDocNamespace) {
+        else if (node.kind === SyntaxKind.Identifier && (<Identifier>node).isInJSDocNamespace) {
             return ModuleInstanceState.NonInstantiated;
         }
         else {
@@ -434,7 +434,11 @@ namespace ts {
                 //       during global merging in the checker. Why? The only case when ambient module is permitted inside another module is module augmentation
                 //       and this case is specially handled. Module augmentations should only be merged with original module definition
                 //       and should never be merged directly with other augmentation, and the latter case would be possible if automatic merge is allowed.
-                if ((!isAmbientModule(node) && (hasExportModifier || container.flags & NodeFlags.ExportContext)) || isJSDocTypedefInJSDocNamespace()) {
+                const isJSDocTypedefInJSDocNamespace = node.kind === SyntaxKind.JSDocTypedefTag && 
+                    node.name && 
+                    node.name.kind === SyntaxKind.Identifier &&
+                    (<Identifier>node.name).isInJSDocNamespace;
+                if ((!isAmbientModule(node) && (hasExportModifier || container.flags & NodeFlags.ExportContext)) || isJSDocTypedefInJSDocNamespace) {
                     const exportKind =
                         (symbolFlags & SymbolFlags.Value ? SymbolFlags.ExportValue : 0) |
                         (symbolFlags & SymbolFlags.Type ? SymbolFlags.ExportType : 0) |
@@ -447,10 +451,6 @@ namespace ts {
                 else {
                     return declareSymbol(container.locals, undefined, node, symbolFlags, symbolExcludes);
                 }
-            }
-
-            function isJSDocTypedefInJSDocNamespace() {
-                return node.kind === SyntaxKind.JSDocTypedefTag && node.name && node.name.flags & NodeFlags.InJSDocNamespace;
             }
         }
 
@@ -1796,8 +1796,10 @@ namespace ts {
             switch (node.kind) {
                 /* Strict mode checks */
                 case SyntaxKind.Identifier:
-                    // for typedef type names with namespaces
-                    if (node.flags & NodeFlags.InJSDocNamespace) {
+                    // for typedef type names with namespaces, bind the new jsdoc type symbol here
+                    // because it requires all containing namespaces to be in effect, namely the
+                    // current "blockScopeContainer" needs to be set to its immediate namespace parent. 
+                    if ((<Identifier>node).isInJSDocNamespace) {
                         let parentNode = node.parent;
                         while (parentNode && parentNode.kind !== SyntaxKind.JSDocTypedefTag) {
                             parentNode = parentNode.parent;

--- a/src/compiler/binder.ts
+++ b/src/compiler/binder.ts
@@ -434,8 +434,8 @@ namespace ts {
                 //       during global merging in the checker. Why? The only case when ambient module is permitted inside another module is module augmentation
                 //       and this case is specially handled. Module augmentations should only be merged with original module definition
                 //       and should never be merged directly with other augmentation, and the latter case would be possible if automatic merge is allowed.
-                const isJSDocTypedefInJSDocNamespace = node.kind === SyntaxKind.JSDocTypedefTag && 
-                    node.name && 
+                const isJSDocTypedefInJSDocNamespace = node.kind === SyntaxKind.JSDocTypedefTag &&
+                    node.name &&
                     node.name.kind === SyntaxKind.Identifier &&
                     (<Identifier>node.name).isInJSDocNamespace;
                 if ((!isAmbientModule(node) && (hasExportModifier || container.flags & NodeFlags.ExportContext)) || isJSDocTypedefInJSDocNamespace) {
@@ -1798,7 +1798,7 @@ namespace ts {
                 case SyntaxKind.Identifier:
                     // for typedef type names with namespaces, bind the new jsdoc type symbol here
                     // because it requires all containing namespaces to be in effect, namely the
-                    // current "blockScopeContainer" needs to be set to its immediate namespace parent. 
+                    // current "blockScopeContainer" needs to be set to its immediate namespace parent.
                     if ((<Identifier>node).isInJSDocNamespace) {
                         let parentNode = node.parent;
                         while (parentNode && parentNode.kind !== SyntaxKind.JSDocTypedefTag) {

--- a/src/compiler/binder.ts
+++ b/src/compiler/binder.ts
@@ -54,6 +54,11 @@ namespace ts {
             const body = (<ModuleDeclaration>node).body;
             return body ? getModuleInstanceState(body) : ModuleInstanceState.Instantiated;
         }
+        // Only jsdoc typedef definition can exist in jsdoc namespace, and it should
+        // be considered the same as type alias
+        else if (node.kind === SyntaxKind.Identifier && node.flags & NodeFlags.InJSDocNamespace) {
+            return ModuleInstanceState.NonInstantiated;
+        }
         else {
             return ModuleInstanceState.Instantiated;
         }
@@ -429,7 +434,7 @@ namespace ts {
                 //       during global merging in the checker. Why? The only case when ambient module is permitted inside another module is module augmentation
                 //       and this case is specially handled. Module augmentations should only be merged with original module definition
                 //       and should never be merged directly with other augmentation, and the latter case would be possible if automatic merge is allowed.
-                if ((!isAmbientModule(node) && (hasExportModifier || container.flags & NodeFlags.ExportContext)) || isJSDocTypedefNamespaceMember()) {
+                if ((!isAmbientModule(node) && (hasExportModifier || container.flags & NodeFlags.ExportContext)) || isJSDocTypedefInJSDocNamespace()) {
                     const exportKind =
                         (symbolFlags & SymbolFlags.Value ? SymbolFlags.ExportValue : 0) |
                         (symbolFlags & SymbolFlags.Type ? SymbolFlags.ExportType : 0) |
@@ -444,11 +449,8 @@ namespace ts {
                 }
             }
 
-            function isJSDocTypedefNamespaceMember() {
-                return node.kind === SyntaxKind.JSDocTypedefTag &&
-                    container.kind === SyntaxKind.ModuleDeclaration &&
-                    (<JSDocTypedefTag>node).name &&
-                    container === (<JSDocTypedefTag>node).name.parent;
+            function isJSDocTypedefInJSDocNamespace() {
+                return node.kind === SyntaxKind.JSDocTypedefTag && node.name && node.name.flags & NodeFlags.InJSDocNamespace;
             }
         }
 
@@ -1795,17 +1797,14 @@ namespace ts {
                 /* Strict mode checks */
                 case SyntaxKind.Identifier:
                     // for typedef type names with namespaces
-                    if (isInJavaScriptFile(node) && node.parent && node.parent.kind === SyntaxKind.ModuleDeclaration) {
+                    if (node.flags & NodeFlags.InJSDocNamespace) {
                         let parentNode = node.parent;
                         while (parentNode && parentNode.kind !== SyntaxKind.JSDocTypedefTag) {
                             parentNode = parentNode.parent;
                         }
-                        if (parentNode && parentNode.kind === SyntaxKind.JSDocTypedefTag && (<JSDocTypedefTag>parentNode).name === node) {
-                            bindBlockScopedDeclaration(<Declaration>parentNode, SymbolFlags.TypeAlias, SymbolFlags.TypeAliasExcludes);
-                            break;
-                        }
+                        bindBlockScopedDeclaration(<Declaration>parentNode, SymbolFlags.TypeAlias, SymbolFlags.TypeAliasExcludes);
+                        break;
                     }
-
                 case SyntaxKind.ThisKeyword:
                     if (currentFlow && (isExpression(node) || parent.kind === SyntaxKind.ShorthandPropertyAssignment)) {
                         node.flowNode = currentFlow;

--- a/src/compiler/parser.ts
+++ b/src/compiler/parser.ts
@@ -411,6 +411,7 @@ namespace ts {
                 return visitNodes(cbNodes, (<JSDocTemplateTag>node).typeParameters);
             case SyntaxKind.JSDocTypedefTag:
                 return visitNode(cbNode, (<JSDocTypedefTag>node).typeExpression) ||
+                    visitNode(cbNode, (<JSDocTypedefTag>node).fullName) ||
                     visitNode(cbNode, (<JSDocTypedefTag>node).name) ||
                     visitNode(cbNode, (<JSDocTypedefTag>node).jsDocTypeLiteral);
             case SyntaxKind.JSDocTypeLiteral:
@@ -6552,7 +6553,14 @@ namespace ts {
                     const typedefTag = <JSDocTypedefTag>createNode(SyntaxKind.JSDocTypedefTag, atToken.pos);
                     typedefTag.atToken = atToken;
                     typedefTag.tagName = tagName;
-                    typedefTag.name = parseJSDocIdentifierName();
+                    typedefTag.fullName = parseJSDocTypeNameWithNamespace(/*flags*/ 0);
+                    if (typedefTag.fullName) {
+                        let rightNode = typedefTag.fullName;
+                        while (rightNode.kind !== SyntaxKind.Identifier) {
+                            rightNode = rightNode.body;
+                        }
+                        typedefTag.name = rightNode;
+                    }
                     typedefTag.typeExpression = typeExpression;
                     skipWhitespace();
 
@@ -6614,6 +6622,22 @@ namespace ts {
                         }
                         scanner.setTextPos(resumePos);
                         return finishNode(jsDocTypeLiteral);
+                    }
+
+                    function parseJSDocTypeNameWithNamespace(flags: NodeFlags) {
+                        const pos = scanner.getTokenPos();
+                        const typeNameOrNamespaceName = parseJSDocIdentifierName();
+
+                        if (typeNameOrNamespaceName && parseOptional(SyntaxKind.DotToken)) {
+                            const jsDocNamespaceNode = <JSDocNamespaceDeclaration>createNode(SyntaxKind.ModuleDeclaration, pos);
+                            jsDocNamespaceNode.flags |= flags;
+                            jsDocNamespaceNode.name = typeNameOrNamespaceName;
+                            jsDocNamespaceNode.body = parseJSDocTypeNameWithNamespace(NodeFlags.NestedNamespace | flags);
+                            return jsDocNamespaceNode;
+                        }
+                        else {
+                            return typeNameOrNamespaceName;
+                        }
                     }
                 }
 

--- a/src/compiler/parser.ts
+++ b/src/compiler/parser.ts
@@ -6637,7 +6637,7 @@ namespace ts {
                         }
 
                         if (typeNameOrNamespaceName && flags & NodeFlags.NestedNamespace) {
-                            typeNameOrNamespaceName.flags |= NodeFlags.InJSDocNamespace;
+                            typeNameOrNamespaceName.isInJSDocNamespace = true;
                         }
                         return typeNameOrNamespaceName;
                     }

--- a/src/compiler/parser.ts
+++ b/src/compiler/parser.ts
@@ -6632,14 +6632,17 @@ namespace ts {
                             const jsDocNamespaceNode = <JSDocNamespaceDeclaration>createNode(SyntaxKind.ModuleDeclaration, pos);
                             jsDocNamespaceNode.flags |= flags;
                             jsDocNamespaceNode.name = typeNameOrNamespaceName;
-                            jsDocNamespaceNode.body = parseJSDocTypeNameWithNamespace(NodeFlags.NestedNamespace | flags);
+                            jsDocNamespaceNode.body = parseJSDocTypeNameWithNamespace(NodeFlags.NestedNamespace);
                             return jsDocNamespaceNode;
                         }
-                        else {
-                            return typeNameOrNamespaceName;
+
+                        if (typeNameOrNamespaceName && flags & NodeFlags.NestedNamespace) {
+                            typeNameOrNamespaceName.flags |= NodeFlags.InJSDocNamespace;
                         }
+                        return typeNameOrNamespaceName;
                     }
                 }
+
 
                 function tryParseChildTag(parentTag: JSDocTypeLiteral): boolean {
                     Debug.assert(token() === SyntaxKind.AtToken);

--- a/src/compiler/scanner.ts
+++ b/src/compiler/scanner.ts
@@ -1799,6 +1799,9 @@ namespace ts {
                 case CharacterCodes.comma:
                     pos++;
                     return token = SyntaxKind.CommaToken;
+                case CharacterCodes.dot:
+                    pos++;
+                    return token = SyntaxKind.DotToken;
             }
 
             if (isIdentifierStart(ch, ScriptTarget.Latest)) {

--- a/src/compiler/types.ts
+++ b/src/compiler/types.ts
@@ -1632,12 +1632,17 @@ namespace ts {
     export interface ModuleDeclaration extends DeclarationStatement {
         kind: SyntaxKind.ModuleDeclaration;
         name: Identifier | LiteralExpression;
-        body?: ModuleBlock | NamespaceDeclaration;
+        body?: ModuleBlock | NamespaceDeclaration | JSDocNamespaceDeclaration | Identifier;
     }
 
     export interface NamespaceDeclaration extends ModuleDeclaration {
         name: Identifier;
         body: ModuleBlock | NamespaceDeclaration;
+    }
+
+    export interface JSDocNamespaceDeclaration extends ModuleDeclaration {
+        name: Identifier;
+        body: JSDocNamespaceDeclaration | Identifier;
     }
 
     export interface ModuleBlock extends Node, Statement {
@@ -1869,6 +1874,7 @@ namespace ts {
 
     export interface JSDocTypedefTag extends JSDocTag, Declaration {
         kind: SyntaxKind.JSDocTypedefTag;
+        fullName?: JSDocNamespaceDeclaration | Identifier;
         name?: Identifier;
         typeExpression?: JSDocTypeExpression;
         jsDocTypeLiteral?: JSDocTypeLiteral;

--- a/src/compiler/types.ts
+++ b/src/compiler/types.ts
@@ -422,6 +422,7 @@ namespace ts {
         JavaScriptFile =     1 << 20, // If node was parsed in a JavaScript
         ThisNodeOrAnySubNodesHasError = 1 << 21, // If this node or any of its children had an error
         HasAggregatedChildData = 1 << 22, // If we've computed data from children and cached it in this node
+        InJSDocNamespace = 1 << 23, // if the node is a member in a JSDoc namespace
 
         BlockScoped = Let | Const,
 

--- a/src/compiler/types.ts
+++ b/src/compiler/types.ts
@@ -421,8 +421,7 @@ namespace ts {
         ThisNodeHasError =   1 << 19, // If the parser encountered an error when parsing the code that created this node
         JavaScriptFile =     1 << 20, // If node was parsed in a JavaScript
         ThisNodeOrAnySubNodesHasError = 1 << 21, // If this node or any of its children had an error
-        HasAggregatedChildData = 1 << 22, // If we've computed data from children and cached it in this node
-        InJSDocNamespace = 1 << 23, // if the node is a member in a JSDoc namespace
+        HasAggregatedChildData = 1 << 22, // If we've computed data from children and cached it in this node 
 
         BlockScoped = Let | Const,
 
@@ -544,6 +543,7 @@ namespace ts {
         originalKeywordKind?: SyntaxKind;              // Original syntaxKind which get set so that we can report an error later
         /*@internal*/ autoGenerateKind?: GeneratedIdentifierKind; // Specifies whether to auto-generate the text for an identifier.
         /*@internal*/ autoGenerateId?: number;         // Ensures unique generated identifiers get unique names, but clones get the same name.
+        isInJSDocNamespace?: boolean;                  // if the node is a member in a JSDoc namespace
     }
 
     // Transient identifier node (marked by id === -1)

--- a/src/compiler/utilities.ts
+++ b/src/compiler/utilities.ts
@@ -3034,7 +3034,7 @@ namespace ts {
             }
         }
 
-        if (node.flags & NodeFlags.NestedNamespace) {
+        if (node.flags & (NodeFlags.NestedNamespace | NodeFlags.InJSDocNamespace)) {
             flags |= ModifierFlags.Export;
         }
 

--- a/src/compiler/utilities.ts
+++ b/src/compiler/utilities.ts
@@ -3034,7 +3034,7 @@ namespace ts {
             }
         }
 
-        if (node.flags & (NodeFlags.NestedNamespace | NodeFlags.InJSDocNamespace)) {
+        if (node.flags & NodeFlags.NestedNamespace || (node.kind === SyntaxKind.Identifier && (<Identifier>node).isInJSDocNamespace)) {
             flags |= ModifierFlags.Export;
         }
 

--- a/tests/baselines/reference/JSDocParsing/DocComments.parsesCorrectly.typedefTagWithChildrenTags.json
+++ b/tests/baselines/reference/JSDocParsing/DocComments.parsesCorrectly.typedefTagWithChildrenTags.json
@@ -18,6 +18,12 @@
                 "end": 16,
                 "text": "typedef"
             },
+            "fullName": {
+                "kind": "Identifier",
+                "pos": 17,
+                "end": 23,
+                "text": "People"
+            },
             "name": {
                 "kind": "Identifier",
                 "pos": 17,

--- a/tests/cases/fourslash/server/jsdocTypedefTagNamespace.ts
+++ b/tests/cases/fourslash/server/jsdocTypedefTagNamespace.ts
@@ -1,0 +1,27 @@
+/// <reference path="../fourslash.ts"/>
+
+// @allowNonTsExtensions: true
+// @Filename: jsdocCompletion_typedef.js
+
+//// /**
+////  * @typedef {string | number} T.NumberLike
+////  * @typedef {{age: number}} T.People
+////  * @typedef {string | number} T.O.Q.NumberLike
+////  * @type {T.NumberLike}
+////  */
+//// var x; x./*1*/;
+//// /** @type {T.O.Q.NumberLike} */
+//// var x1; x1./*2*/;
+//// /** @type {T.People} */
+//// var x1; x1./*3*/;
+
+goTo.marker("1");
+verify.memberListContains('charAt');
+verify.memberListContains('toExponential');
+
+goTo.marker("2");
+verify.memberListContains('age');
+
+goTo.marker("3");
+verify.memberListContains('charAt');
+verify.memberListContains('toExponential');

